File: docs/breez-sdk/snippets/go/cross_chain.go
package example

import (
	"errors"
	"log"
	"math/big"

	"github.com/breez/breez-sdk-spark-go/breez_sdk_spark"
)

func GetCrossChainRoutes(sdk *breez_sdk_spark.BreezSdk) ([]breez_sdk_spark.CrossChainRoutePair, error) {
	// ANCHOR: cross-chain-get-routes
	inputStr := "<recipient address>"
	input, err := sdk.Parse(inputStr)
	if err != nil {
		var sdkErr *breez_sdk_spark.SdkError
		if errors.As(err, &sdkErr) {
			// Handle SdkError
		}
		return nil, err
	}

	addressInput, ok := input.(breez_sdk_spark.InputTypeCrossChainAddress)
	if !ok {
		return nil, errors.New("not a cross-chain address")
	}
	addressDetails := addressInput.Field0

	filter := breez_sdk_spark.CrossChainRouteFilterSend{AddressDetails: addressDetails}
	routes, err := sdk.GetCrossChainRoutes(filter)
	if err != nil {
		return nil, err
	}

	for _, route := range routes {
		log.Printf("Route via %v: %s/%s", route.Provider, route.Chain, route.Asset)
		// Amount bounds are published per accepted asset. Read the ones for
		// the asset you intend to pay with, before quoting.
		for _, accepted := range route.AcceptedAssets {
			limits := accepted.Limits
			if limits == nil {
				continue
			}
			if limits.MinAmount != nil {
				log.Printf("  %v minimum: %v base units", accepted.Asset, *limits.MinAmount)
			}
			if limits.MinUsdCents != nil {
				log.Printf("  %v minimum: %d USD cents", accepted.Asset, *limits.MinUsdCents)
			}
			if limits.DynamicLimitsPossible {
				log.Printf("  The provider may enforce a higher minimum than published")
			}
		}
	}
	// ANCHOR_END: cross-chain-get-routes
	return routes, nil
}

func PrepareSendPaymentCrossChain(
	sdk *breez_sdk_spark.BreezSdk,
	addressDetails breez_sdk_spark.CrossChainAddressDetails,
	route breez_sdk_spark.CrossChainRoutePair,
) (*breez_sdk_spark.PrepareSendPaymentResponse, error) {
	// ANCHOR: cross-chain-prepare
	// Optionally set the maximum slippage in basis points (10 to 500)
	optionalMaxSlippageBps := uint32(100)
	amount := new(big.Int).SetInt64(50_000)

	request := breez_sdk_spark.PrepareSendPaymentRequest{
		PaymentRequest: breez_sdk_spark.PaymentRequestCrossChain{
			Address:           addressDetails.Address,
			Route:             route,
			MaxSlippageBps:    &optionalMaxSlippageBps,
			TargetOverpayBps:  nil,
		},
		Amount:            &amount,
		TokenIdentifier:   nil,
		ConversionOptions: nil,
		FeePolicy:         nil,
	}
	response, err := sdk.PrepareSendPayment(request)
	if err != nil {
		return nil, err
	}

	switch paymentMethod := response.PaymentMethod.(type) {
	case breez_sdk_spark.SendPaymentMethodCrossChainAddress:
		log.Printf("Amount in: %v", paymentMethod.AmountIn)
		log.Printf("Estimated out: %v", paymentMethod.EstimatedOut)
		log.Printf("Provider fee: %v", paymentMethod.FeeAmount)
		log.Printf("Quote expires at: %s", paymentMethod.ExpiresAt)
	}
	// ANCHOR_END: cross-chain-prepare
	return &response, nil
}

func SendPaymentCrossChain(
	sdk *breez_sdk_spark.BreezSdk,
	prepareResponse breez_sdk_spark.PrepareSendPaymentResponse,
) (*breez_sdk_spark.SendPaymentResponse, error) {
	// ANCHOR: cross-chain-send
	// Only valid for sends with no token leg (see Retry safety).
	optionalIdempotencyKey := "<idempotency key uuid>"
	request := breez_sdk_spark.SendPaymentRequest{
		PrepareResponse: prepareResponse,
		Options:         nil,
		IdempotencyKey:  &optionalIdempotencyKey,
	}
	response, err := sdk.SendPayment(request)
	if err != nil {
		return nil, err
	}
	log.Printf("Payment: %v", response.Payment)
	// ANCHOR_END: cross-chain-send
	return &response, nil
}

func GetCrossChainReceiveRoutes(sdk *breez_sdk_spark.BreezSdk) ([]breez_sdk_spark.CrossChainRoutePair, error) {
	// ANCHOR: cross-chain-get-receive-routes
	filter := breez_sdk_spark.CrossChainRouteFilterReceive{ContractAddress: nil}
	routes, err := sdk.GetCrossChainRoutes(filter)
	if err != nil {
		return nil, err
	}

	for _, route := range routes {
		log.Printf(
			"Route via %v: %s/%s -> Spark",
			route.Provider, route.Chain, route.Asset,
		)
	}
	// ANCHOR_END: cross-chain-get-receive-routes
	return routes, nil
}

func ReceivePaymentCrossChain(
	sdk *breez_sdk_spark.BreezSdk,
	route breez_sdk_spark.CrossChainRoutePair,
) (*breez_sdk_spark.ReceivePaymentResponse, error) {
	// ANCHOR: cross-chain-receive
	// amount is in the route's source-asset base units (USD-stable parity:
	// 1_000_000 = $1 on 6-decimal routes). See the guide for FeeMode,
	// destination, and the slippage/overpay overrides.
	amount := new(big.Int).SetInt64(1_000_000)
	var optionalDestination *breez_sdk_spark.SparkAsset = nil
	optionalMaxSlippageBps := uint32(100)
	var optionalTargetOverpayBps *uint32 = nil
	var optionalFeeMode *breez_sdk_spark.CrossChainFeeMode = nil

	request := breez_sdk_spark.ReceivePaymentRequest{
		PaymentMethod: breez_sdk_spark.ReceivePaymentMethodCrossChain{
			Route:             route,
			Amount:            amount,
			Destination:       optionalDestination,
			FeeMode:           optionalFeeMode,
			MaxSlippageBps:    &optionalMaxSlippageBps,
			TargetOverpayBps:  optionalTargetOverpayBps,
		},
	}
	response, err := sdk.ReceivePayment(request)
	if err != nil {
		return nil, err
	}

	log.Printf("Payment request: %s", response.PaymentRequest)
	if info := response.CrossChainInfo; info != nil {
		log.Printf("Deposit address: %s", info.DepositAddress)
		log.Printf("Deposit amount: %v", info.DepositAmount)
		log.Printf(
			"Expected received: %v %s",
			info.ExpectedReceivedAmount, info.DestinationAsset,
		)
		log.Printf("Expires at: %d", info.ExpiresAt)
	}
	// ANCHOR_END: cross-chain-receive
	return &response, nil
}

File: docs/breez-sdk/snippets/rust/src/cross_chain.rs
use anyhow::Result;
use breez_sdk_spark::*;
use log::info;

async fn get_cross_chain_routes(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: cross-chain-get-routes
    let input = "<recipient address>";
    let InputType::CrossChainAddress(address_details) = sdk.parse(input).await? else {
        anyhow::bail!("Not a cross-chain address");
    };

    let routes = sdk
        .get_cross_chain_routes(&CrossChainRouteFilter::Send {
            address_details: address_details.clone(),
        })
        .await?;

    for route in &routes {
        info!(
            "Route via {:?}: {}/{}",
            route.provider, route.chain, route.asset
        );
        // Amount bounds are published per accepted asset. Read the ones for
        // the asset you intend to pay with, before quoting.
        for accepted in &route.accepted_assets {
            let Some(limits) = &accepted.limits else {
                continue;
            };
            info!(
                "  {:?} minimum: {:?} base units / {:?} USD cents",
                accepted.asset, limits.min_amount, limits.min_usd_cents
            );
            if limits.dynamic_limits_possible {
                info!("  The provider may enforce a higher minimum than published");
            }
        }
    }
    // ANCHOR_END: cross-chain-get-routes
    Ok(())
}

async fn prepare_send_payment_cross_chain(
    sdk: &BreezSdk,
    address_details: CrossChainAddressDetails,
    route: CrossChainRoutePair,
) -> Result<()> {
    // ANCHOR: cross-chain-prepare
    // Optionally set the maximum slippage in basis points (10 to 500)
    let optional_max_slippage_bps = Some(100);

    let prepare_response = sdk
        .prepare_send_payment(PrepareSendPaymentRequest {
            payment_request: PaymentRequest::CrossChain {
                address: address_details.address.clone(),
                route,
                max_slippage_bps: optional_max_slippage_bps,
                target_overpay_bps: None,
            },
            amount: Some(50_000),
            token_identifier: None,
            conversion_options: None,
            fee_policy: None,
        })
        .await?;

    if let SendPaymentMethod::CrossChainAddress {
        amount_in,
        estimated_out,
        fee_amount,
        expires_at,
        ..
    } = &prepare_response.payment_method
    {
        info!("Amount in: {amount_in}");
        info!("Estimated out: {estimated_out}");
        info!("Provider fee: {fee_amount}");
        info!("Quote expires at: {expires_at}");
    }
    // ANCHOR_END: cross-chain-prepare
    Ok(())
}

async fn send_payment_cross_chain(
    sdk: &BreezSdk,
    prepare_response: PrepareSendPaymentResponse,
) -> Result<()> {
    // ANCHOR: cross-chain-send
    // Only valid for sends with no token leg (see Retry safety).
    let optional_idempotency_key = Some("<idempotency key uuid>".to_string());
    let send_response = sdk
        .send_payment(SendPaymentRequest {
            prepare_response,
            options: None,
            idempotency_key: optional_idempotency_key,
        })
        .await?;
    let payment = send_response.payment;
    info!("Payment: {payment:?}");
    // ANCHOR_END: cross-chain-send
    Ok(())
}

async fn get_cross_chain_receive_routes(sdk: &BreezSdk) -> Result<()> {
    // ANCHOR: cross-chain-get-receive-routes
    let routes = sdk
        .get_cross_chain_routes(&CrossChainRouteFilter::Receive {
            contract_address: None,
        })
        .await?;

    for route in &routes {
        info!(
            "Route via {:?}: {}/{} -> Spark",
            route.provider, route.chain, route.asset
        );
    }
    // ANCHOR_END: cross-chain-get-receive-routes
    Ok(())
}

async fn receive_payment_cross_chain(sdk: &BreezSdk, route: CrossChainRoutePair) -> Result<()> {
    // ANCHOR: cross-chain-receive
    // amount is in the route's source-asset base units (USD-stable parity:
    // 1_000_000 = $1 on 6-decimal routes). See the guide for fee_mode,
    // destination, and the slippage/overpay overrides.
    let amount = 1_000_000u128;
    let optional_destination: Option<SparkAsset> = None;
    let optional_max_slippage_bps = Some(100);
    let optional_target_overpay_bps: Option<u32> = None;
    let optional_fee_mode: Option<CrossChainFeeMode> = None;

    let response = sdk
        .receive_payment(ReceivePaymentRequest {
            payment_method: ReceivePaymentMethod::CrossChain {
                route,
                amount,
                destination: optional_destination,
                fee_mode: optional_fee_mode,
                max_slippage_bps: optional_max_slippage_bps,
                target_overpay_bps: optional_target_overpay_bps,
            },
        })
        .await?;

    info!("Payment request: {}", response.payment_request);
    if let Some(info) = response.cross_chain_info {
        info!("Deposit address: {}", info.deposit_address);
        info!("Deposit amount: {}", info.deposit_amount);
        info!(
            "Expected received: {} {}",
            info.expected_received_amount, info.destination_asset
        );
        info!("Expires at: {}", info.expires_at);
    }
    // ANCHOR_END: cross-chain-receive
    Ok(())
}
